### DISCUSSION TITLE
GPUI with UI Unit Test

### DISCUSSION THREAD

**Discussion by qinyuhang** [zed-industries/zed#17477] [Feature Requests]
### GPUI with UI Unit Test
Hi, I'm interested in GPUI crate, I'm wondering if we have any method to perform UI Unit Test to check the correctness of code utilized GPUI. Just Like Jest+React-Testing-Library for Javascript or Xcode UI Unit test for Swift.

Eg: 
1. Snapshot Test: Giving certain State, dump the UI to xml/html, then check the xml content matches certain snapshot string.
2. Reaction Test: Giving certain State, dispatch certain action on certain element, then check if desired reaction/action were trigged.

**Comment by MatinAniss** [comment]
GPUI does have the macro `gpui::test` which allows you to simulate keystrokes, change state, and more. I think this is what you are looking for.

You can see it documented here https://github.com/zed-industries/zed/blob/main/crates/gpui_macros/src/gpui_macros.rs#L85C1

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
